docs/proposals/generalize-modules.md

```diff
@@ -0,0 +1,66 @@
+---
+title: "Generalize Modules Service to make it extensible"
+linkTitle: "Generalize Modules Service to make it extensible"
+weight: 1
+slug: generalize-modules
+---
+
+- Author: @annanay25
+- Reviewers: @jtlisi, @pstibrany
+- Date: April 2020
+- Status: Review
+
+## Overview
+
+Cortex uses modules to start and operate services with dependencies. Inter-service dependencies are specified in a map and passed to a module manager which ensures that they are started in the right order of dependencies. While this works really well, the implementation is tied in specifically to the Cortex struct and is not flexible for use with other projects like Loki, which also require similar forms of dependency management.
+
+We would like to extend modules in cortex to a generic dependency management framework, that can be used by any project with no ties to cortex.
+
+## Specific goals
+
+- Framework should allow for reusing cortex modules and allow us to:
+  - Add new modules
+  - Overwrite the implementation of a current module
+  - Manage dependencies
+- Framework should allow for building an application from scratch using the `modules` package, with no dependencies on Cortex. For ex: Remove code from Loki that was copied from `pkg/cortex/cortex.go`.
+
+
+## Proposed Design
+
+### Modules package
+
+To make the modules package extensible, we need to abstract away any Cortex specific details from the module manager. The proposed design is to -
+
+- Make a new component `Manager`, which is envisioned to be a central manager for all modules of the application. It stores modules & dependencies, and will be housed under a new package `pkg/util/modules`. `Manager` has the following methods for interaction:
+```
+   func (m *Manager) RegisterModule(name string, deps []string, initFn func() (Service, error))
+   func (m *Manager) AddDependency(fromModule string, toModule string) error
+   func (m *Manager) InitModuleServices(target string) (map[string]services.Service, error)
```

Comment:
Have you considered calling it `Start()` given I assume it will start module services?

Reply:
We should mention ability to pass multiple targets as a future extension (unless you want to put it into first implementation, but that's not necessary).

Reply:
It initialises all the module services, but does not start them. I would prefer calling it `InitModuleServices`.

> We should mention ability to pass multiple targets as a future extension

Definitely. I will implement it as a follow up after the first implementation is merged.

Reply:
Agree with Annanay, this naming is clearer. Method documentation should be clear about returning services in New state. Also requirement of init function is that returned service (if not nil) is in New state -- module.Manager can test that and return error if that's not true.

Reply:
Agree too. I initially misunderstood it. I thought it was also starting them, not just initialising. All good.